packages/xrpc-server/src/stream/frames.ts

```diff
@@ -0,0 +1,122 @@
+import * as cborx from 'cbor-x'
+import * as uint8arrays from 'uint8arrays'
+import {
+  frameHeader,
+  FrameHeader,
+  FrameType,
+  InfoFrameHeader,
+  DataFrameHeader,
+  ErrorFrameHeader,
+  infoFrameBody,
+  InfoFrameBody,
+  ErrorFrameBody,
+  errorFrameBody,
+} from './types'
+
+export abstract class Frame {
+  header: FrameHeader
+  body: unknown
+  get op(): FrameType {
+    return this.header.op
+  }
+  toBytes(): Uint8Array {
+    return uint8arrays.concat([
+      cborx.encode(this.header),
+      cborx.encode(this.body),
+    ])
+  }
+  static fromBytes(bytes: Uint8Array) {
+    let i = 0
+    let header: unknown
+    let body: unknown = kUnset
+    cborx.decodeMultiple(bytes, (item) => {
+      if (i === 0) {
+        header = item
+      } else if (i === 1) {
+        body = item
+      } else {
+        throw new Error('Too many CBOR data items in frame')
+      }
+      i++
+    })
+    const parsedHeader = frameHeader.safeParse(header)
+    if (!parsedHeader.success) {
+      throw new Error(`Invalid frame header: ${parsedHeader.error.message}`)
+    }
+    if (body === kUnset) {
+      throw new Error('Missing frame body')
```

Comment:
do we ever have empty frame bodies? or would we send `null` in that case?

Reply:
You would send an empty CBOR data item as the body, which encodes as `f7` in hex.

Reply:
(I needed `kUnset` here since `f7` decodes to `undefined` 😆 )

Reply:
kk that's what i figured 👍